### PR TITLE
feat!: drop legacy storage

### PR DESCRIPTION
**Pull request by benjlevesque**
_BREAKING CHANGE_: Dropping legacy indexing mechanism.

Changes:
- enforce usage of TheGraph as a backend for request-node
- drop all custom indexing code (mostly DataAccess)
- refactor DataAccess to rely on an Indexer
- refactor mock to use in-memory indexer
- reduce dependency to TheGraph by abstracting most of the code transformation. TheGraph dependency is now down to only the SubgraphClient.

### CODE REVIEW

**Comment by MantisClone** [comment]
I recognize that the legacy indexing mechanism doesn't work on many chains, is difficult to maintain, and that no one wants to spend the time to fix it. For those reasons, I agree that removing it now is the right move. 

Nonetheless, would it be possible to leave injection points such that someone could plug in their own custom indexer into the protocol?

**Comment by benjlevesque** [comment]
> I recognize that the legacy indexing mechanism doesn't work on many chains, is difficult to maintain, and that no one wants to spend the time to fix it. For those reasons, I agree that removing it now is the right move.
> 
> Nonetheless, would it be possible to leave injection points such that someone could plug in their own custom indexer into the protocol?

Yes @MantisClone; that would be best indeed! These injection points are still available, but not as the Node level (same as today). I've introduced a new layer called "indexer" that represents better the way things are (it would actually be possible to keep the old code with this indexer layer, but it's not efficient so better remove it). 

This PR is still draft because I'm trying to reduce the surface of TheGraph in the code; it should be down to only the Subgraph Client, graphql queries and types. The rest could be generalised to work with any type of indexer.

File: packages/request-node/src/server.ts
Comment (benjlevesque) [inline packages/request-node/src/server.ts:1]: this file now only creates the server object, but doesn't start it. `startNode` is called from `bin.ts`

File: packages/request-node/src/requestNode.ts
Comment (benjlevesque) [inline packages/request-node/src/requestNode.ts:1]: `requestNodeBase.ts` moved to `requestNode.ts`

File: packages/data-access/src/data-read.ts
Comment (benjlevesque) [inline packages/data-access/src/data-read.ts:1]: moved from `packages/thegraph-data-access/src/data-access.ts`

File: packages/data-access/src/pending-store.ts
Comment (benjlevesque) [inline packages/data-access/src/pending-store.ts:1]: moved from thegraph-data-access

File: packages/smart-contracts/scripts/deploy-payments.ts [hunk 1]
```diff
@@ -10,10 +10,10 @@ import { uniswapV2RouterAddresses, jumpToNonce } from './utils';
 import { Contract } from 'ethers';
 // eslint-disable-next-line
 // @ts-ignore Cannot find module
-import { ChainlinkConversionPath } from '../src/types/ChainlinkConversionPath';
```
Comment (benjlevesque) [inline packages/smart-contracts/scripts/deploy-payments.ts:13]: I had lots of issues due to dependencies (probably because removing some deps changes the resolved version of other packages) and had to upgrade some. This change is related to the typechain upgrade

File: packages/request-node/src/server.ts [hunk 2]
```diff
@@ -1,42 +1,29 @@
-#!/usr/bin/env node
```
Comment (benjlevesque) [inline packages/request-node/src/server.ts:1]: this file now only creates the server object, but doesn't start it. startNode is called from bin.ts

File: packages/data-access/src/data-read.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
+import { DataAccessTypes, StorageTypes } from '@requestnetwork/types';
```
Comment (benjlevesque) [inline packages/data-access/src/data-read.ts:1]: moved from packages/thegraph-data-access/src/data-access.ts

File: packages/data-access/src/pending-store.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { DataAccessTypes } from '@requestnetwork/types';
```
Comment (benjlevesque) [inline packages/data-access/src/pending-store.ts:1]: moved from packages/thegraph-data-access/src/data-access.ts

File: packages/ethereum-storage/src/ethereum-utils.ts [hunk 5]
```diff
@@ -1,8 +1,7 @@
-import { CurrencyTypes, StorageTypes } from '@requestnetwork/types';
+import { StorageTypes } from '@requestnetwork/types';
 import * as config from './config';
 
 import { BigNumber } from 'ethers';
-import { EvmChains } from '@requestnetwork/currency';
```
Comment (benjlevesque) [inline packages/ethereum-storage/src/ethereum-utils.ts:5]: @alexandre-abrioux FYI. 
I'm trying to reduce inter-packages dependencies... The currency package doesn't make sense in the storage I think

File: packages/request-client.js/src/http-metamask-data-access.ts [hunk 6]
```diff
@@ -106,26 +103,26 @@ export default class HttpMetaMaskDataAccess extends HttpDataAccess {
...
-    const tx = await this.submitterContract.submitHash(
+    const tx = await submitterContract.submitHash(
       ipfsHash,
       /* eslint-disable no-magic-numbers */
       ethers.utils.hexZeroPad(ethers.utils.hexlify(ipfsSize), 32),
       { value: fee },
     );
 
-    const ethBlock = await this.provider.getBlock(tx.blockNumber);
+    const ethBlock = await this.provider.getBlock(tx.blockNumber || -1);
```
Comment (yomarion) [inline packages/request-client.js/src/http-metamask-data-access.ts:116]:
```suggestion
    const ethBlock = await this.provider.getBlock(tx.blockNumber ?? -1);
```
Any reason for not sticking with the nullish coalescing? It has impacts in edge cases like blockNumber = 0. Even if unlikely, I thought we should stick with the good practice.

File: packages/request-node/test/getConfirmedTransaction.test.ts [hunk 7]
```diff
@@ -60,7 +65,7 @@ describe('getConfirmedTransaction', () => {
     expect(serverResponse!.body.result).toMatchObject({});
     // 'getConfirmedTransaction request meta'
     expect(serverResponse!.body.meta.storageMeta.state).toBe('confirmed');
-  }, 11000);
+  }, 30000);
```
Comment (yomarion) [inline packages/request-node/test/getConfirmedTransaction.test.ts:68]: Are the 19 additional seconds needed just for the mining?

File: packages/thegraph-data-access/src/subgraph-client.ts [hunk 8]
```diff
@@ -1,51 +1,110 @@
...
     });
   }
 
-  public getChannelsByTopics(topics: string[]): Promise<TransactionsBody> {
-    return this.graphql.request<TransactionsBody>(GetChannelsByTopicsQuery, {
-      topics,
-    });
+  // FIXME: this should be possible to do in a single query to the subgraph,
+  // but currently one transaction doesn't contain topics from previous ones on the same channel.
+  // This could be fixed on the Subgraph indexer code for optimization.
```
Comment (yomarion) [inline packages/thegraph-data-access/src/subgraph-client.ts:53]: 👍 Very clear, thanks for documenting the fixme

**Review by yomarion — APPROVED** [review]
Massive!

File: packages/request-node/test/getConfirmedTransaction.test.ts (hunk 7 above)
Comment (benjlevesque) [reply packages/request-node/test/getConfirmedTransaction.test.ts:68]: indexing of the graph. I increased a bit more than actually needed, to avoid flakiness

File: packages/usage-examples/src/request-client-js-erc20-request.ts [hunk 9]
```diff
@@ -65,7 +65,7 @@ requestNetwork
   .createRequest(createParams)
   .then((request) => {
     console.log('clear request:');
-    console.log(request);
+    console.log(request.requestId);
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/usage-examples/src/request-client-js-erc20-request.ts:68]: I agree with @yomarion - logging the whole request makes sense in usage-example context.

**Review by MantisClone — COMMENTED** [review]
I'm about 1/4th of the way through the review.

File: packages/thegraph-data-access/src/types.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { StorageTypes } from '@requestnetwork/types';
+import { DataAccessBaseOptions } from '@requestnetwork/data-access';
+
+import { RequestInit } from 'graphql-request/dist/types.dom';
```
Comment (MantisClone) [inline packages/thegraph-data-access/src/types.ts:4]: What is this? Why is this imported from dist?

**Review by MantisClone — APPROVED** [review]
Impressive refactor! Only a few minor comments.

File: packages/thegraph-data-access/src/types.ts (hunk 10 above)
Comment (benjlevesque) [reply packages/thegraph-data-access/src/types.ts:4]: because not exported from root :)
(nb: not new, moved)

File: packages/request-client.js/src/http-request-network.ts [hunk 11]
```diff
@@ -9,8 +9,8 @@ import { AxiosRequestConfig } from 'axios';
 import { PaymentNetworkOptions } from '@requestnetwork/payment-detection';
 import RequestNetwork from './api/request-network';
 import HttpDataAccess from './http-data-access';
-import MockDataAccess from './mock-data-access';
-import MockStorage from './mock-storage';
+import { MockDataAccess } from '@requestnetwork/data-access';
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/request-client.js/src/http-request-network.ts:12]: Not sure what you mean, this is already done like this I think? can you see any place where it's not?

The change in this file is due to the Mock Storage not being in the request-client anymore

File: README.md [hunk 12]
```diff
@@ -41,12 +41,11 @@ Join the [Request Discord][request-discord-url] to get in touch with us.
 
 ### Private Packages
 
-| Package                                                                | Description                                                 |
-| ---------------------------------------------------------------------- | ----------------------------------------------------------- |
-| [`@requestnetwork/integration-test`](/packages/integration-test)       | Integration test for the Request system                     |
-| [`@requestnetwork/prototype-estimator`](/packages/prototype-estimator) | Give estimates of size and throughput of the Request system |
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply README.md:47]: I've never seen it used. I'm in favor of dropping unused parts that cause more maintenance than they are used

File: packages/request-client.js/src/http-request-network.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-request-network.ts:12]: I think there are places where things are imported from sibling source files instead of via a package dependency.  I'll find and share an example.

[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-request-network.ts:12]: I went searching and even went so far as to enable an eslint rule [`monorepo/no-relative-import`](https://github.com/azz/eslint-plugin-monorepo#monorepono-relative-import-fixable) but I couldn't find any examples.  I stand corrected 😳

File: packages/request-node/src/dataAccess.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  ipfsStorage: StorageTypes.IIpfsStorage,
+  logger: LogTypes.ILogger,
+): DataAccessTypes.IDataAccess {
+  const graphNodeUrl = config.getGraphNodeUrl();
+
+  const network = getEthereumStorageNetworkNameFromId(config.getStorageNetworkId()) as any;
+  if (!network) {
+    throw new Error(`Storage network not supported: ${config.getStorageNetworkId()}`);
+  }
+  EvmChains.assertChainSupported(network);
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/request-node/src/dataAccess.ts:25]: I agree. I'd like to change this a little bit, in another PR. 
This call is there mostly for type assertion

**Review by alexandre-abrioux — APPROVED** [review]
👏

File: packages/thegraph-data-access/src/subgraph-client.ts (hunk 8 above)
Comment (benjlevesque) [reply packages/thegraph-data-access/src/subgraph-client.ts:53]: Issue https://github.com/RequestNetwork/storage-subgraph/issues/9

**Comment by benjlevesque** [comment]
NB: from now on, the `IPFS_URL` environment replaces `IPFS_HOST`, `IPFS_PORT` and `IPFS_PROTOCOL`
